### DISCUSSION TITLE
Refactor procedural code into functions from the Copilot Labs panel

### DISCUSSION THREAD

**Discussion by ozkary** [community/community#67627] [Copilot Conversations]
### Select Topic Area

Product Feedback

## Provide the ability to refactor code into functions

When some procedure code is refactored, we usually select the code area and refactor it into one or more functions, depending on the size and logic of the code. It would be great to be able to select certain lines of code and be able to open Copilot Labs, and ask it to refactor into a function. 

- This feature should work similar to code translation where a preview of the function can be provided
- Copilot should infer the parameters from the code variables
- Copilot should infer and suggest the return type

This should be just a tool to create the function template with the code, and allow the developer to refine the refactor code, parameters, etc.

### This is a simple example for illustration purposes. Real code samples can contain 50+ lines

```csharp
// procedural code
var unitsSold = 100;
var unitPrice = 2;
int totalPrice= unitsSold * unitPrice;

```
### Copilot Labs Result

```csharp
//refactored code in the same or new class
int CalculateTotalPrice(int units, int price) {
  return unit * price;
}

```
### The previous lines are then replaced by

```csharp
// new code
int totalPrice= CalculateTotalPrice(unitsSold, unitPrice);
```

**Comment by Akash1134** [comment]
Hi @ozkary , Thank you for sharing your feedback. We’d love to get your help in making GitHub Copilot Labs better! You may also reach out to us on [Discord](https://gh.io/next-discord) to share your feedback.
